File: crates/stats-code/src/math/distributions.rs
//! Named distribution helpers for statistical methods.

#![allow(dead_code)]

use super::{
    chi_square_cdf, f_distribution_p_value, t_distribution_critical_value, t_distribution_p_value,
};

/// Two-sided upper-tail p-value for Student's t statistic.
pub(crate) fn student_t_two_sided(t: f64, df: f64) -> f64 {
    t_distribution_p_value(t, df)
}

/// Positive two-sided Student's t critical value for significance level alpha.
pub(crate) fn student_t_inv(alpha: f64, df: f64) -> f64 {
    t_distribution_critical_value(alpha, df)
}

/// Upper-tail p-value for an F statistic.
pub(crate) fn f_distribution_p(f: f64, df1: f64, df2: f64) -> f64 {
    f_distribution_p_value(f, df1, df2)
}

/// Upper-tail p-value for a chi-square statistic.
pub(crate) fn chi_square_p(x: f64, df: f64) -> f64 {
    if x <= 0.0 || df <= 0.0 {
        return 1.0;
    }
    (1.0 - chi_square_cdf(x, df)).clamp(0.0, 1.0)
}

/// Approximate upper-tail p-value for Tukey's studentized range statistic.
///
/// This deterministic approximation uses the relationship `q / sqrt(2) ~ t`
/// for each pairwise contrast and applies a Sidak-style max-tail adjustment
/// across `k * (k - 1) / 2` comparisons.
#[must_use]
pub fn studentized_range_p(q: f64, k: usize, df: f64) -> f64 {
    if !q.is_finite() || q <= 0.0 || k < 2 || df <= 0.0 {
        return 1.0;
    }
    let comparisons = (k * (k - 1) / 2) as f64;
    let pair_tail = student_t_two_sided(q / 2.0_f64.sqrt(), df).clamp(0.0, 1.0);
    (1.0 - (1.0 - pair_tail).powf(comparisons)).clamp(0.0, 1.0)
}

/// Approximate Lilliefors-corrected K-S p-value for normality diagnostics.
pub(crate) fn lilliefors_p(d: f64, n: usize) -> f64 {
    if !d.is_finite() || d <= 0.0 || n == 0 {
        return 1.0;
    }
    let n = n as f64;
    let z = (n.sqrt() - 0.01 + 0.85 / n.sqrt()) * d;
    if z < 0.302 {
        1.0
    } else if z < 1.18 {
        (1.093 - 7.01256 * z + 2.99587 * z * z).exp()
    } else {
        (1.0776 - 5.22323 * z).exp()
    }
    .clamp(0.0, 1.0)
}

/// Inverse standard normal CDF using Peter J. Acklam's rational approximation.
pub(crate) fn inverse_normal(p: f64) -> f64 {
    if p <= 0.0 {
        return f64::NEG_INFINITY;
    }
    if p >= 1.0 {
        return f64::INFINITY;
    }

    const A: [f64; 6] = [
        -3.969_683_028_665_376e1,
        2.209_460_984_245_205e2,
        -2.759_285_104_469_687e2,
        1.383_577_518_672_69e2,
        -3.066_479_806_614_716e1,
        2.506_628_277_459_239,
    ];
    const B: [f64; 5] = [
        -5.447_609_879_822_406e1,
        1.615_858_368_580_409e2,
        -1.556_989_798_598_866e2,
        6.680_131_188_771_972e1,
        -1.328_068_155_288_572e1,
    ];
    const C: [f64; 6] = [
        -7.784_894_002_430_293e-3,
        -3.223_964_580_411_365e-1,
        -2.400_758_277_161_838,
        -2.549_732_539_343_734,
        4.374_664_141_464_968,
        2.938_163_982_698_783,
    ];
    const D: [f64; 4] = [
        7.784_695_709_041_462e-3,
        3.224_671_290_700_398e-1,
        2.445_134_137_142_996,
        3.754_408_661_907_416,
    ];

    let plow = 0.02425;
    let phigh = 1.0 - plow;
    if p < plow {
        let q = (-2.0 * p.ln()).sqrt();
        return (((((C[0] * q + C[1]) * q + C[2]) * q + C[3]) * q + C[4]) * q + C[5])
            / ((((D[0] * q + D[1]) * q + D[2]) * q + D[3]) * q + 1.0);
    }
    if p > phigh {
        let q = (-2.0 * (1.0 - p).ln()).sqrt();
        return -(((((C[0] * q + C[1]) * q + C[2]) * q + C[3]) * q + C[4]) * q + C[5])
            / ((((D[0] * q + D[1]) * q + D[2]) * q + D[3]) * q + 1.0);
    }
    let q = p - 0.5;
    let r = q * q;
    (((((A[0] * r + A[1]) * r + A[2]) * r + A[3]) * r + A[4]) * r + A[5]) * q
        / (((((B[0] * r + B[1]) * r + B[2]) * r + B[3]) * r + B[4]) * r + 1.0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::math::normal_cdf;

    #[test]
    fn named_distribution_helpers_match_existing_kernels() {
        assert!((student_t_two_sided(2.0, 20.0) - t_distribution_p_value(2.0, 20.0)).abs() < 1e-12);
        assert!(
            (f_distribution_p(3.0, 2.0, 20.0) - f_distribution_p_value(3.0, 2.0, 20.0)).abs()
                < 1e-12
        );
        assert!((chi_square_p(3.841, 1.0) - 0.05).abs() < 0.01);
    }

    #[test]
    fn inverse_normal_known_values() {
        assert!(inverse_normal(0.5).abs() < 1e-12);
        assert!((normal_cdf(inverse_normal(0.975)) - 0.975).abs() < 1e-6);
    }

    #[test]
    fn adjusted_p_values_are_probabilities() {
        assert!((0.0..=1.0).contains(&studentized_range_p(3.0, 4, 20.0)));
        assert!((0.0..=1.0).contains(&lilliefors_p(0.12, 30)));
    }
}

File: ts-backend/packages/engine/src/math/distributions.ts
// math/distributions.ts — probability distribution CDFs and p-values (task 5.1).
// Transcribed from crates/stats-code/src/math/{mod,distributions}.rs.

import {
  regularizedLowerGamma,
  regularizedUpperGammaCf,
  regularizedBetaIncomplete,
  regularizedIncompleteBeta,
  logGamma,
  clamp01,
  SQRT_2PI,
} from './special.js';

/** Standard normal CDF (Abramowitz & Stegun approximation, ~1e-7 precision). */
export function normalCdf(value: number): number {
  const absolute = Math.abs(value);
  const t = 1 / (1 + 0.2316419 * absolute);
  const density = Math.exp(-0.5 * absolute * absolute) / SQRT_2PI;
  const approximation =
    1 -
    density *
      t *
      (0.31938153 +
        t * (-0.356563782 + t * (1.781477937 + t * (-1.821255978 + t * 1.330274429))));
  return value >= 0 ? approximation : 1 - approximation;
}

/** Chi-square CDF using the regularized lower incomplete gamma function. */
export function chiSquareCdf(x: number, df: number): number {
  if (x <= 0 || df <= 0) return 0;
  return regularizedLowerGamma(df / 2, x / 2);
}

/** Upper-tail chi-square p-value. */
export function chiSquareP(x: number, df: number): number {
  if (x <= 0 || df <= 0) return 1;
  return clamp01(1 - chiSquareCdf(x, df));
}

/** Upper-tail F-distribution p-value via the incomplete beta relationship. */
export function fDistributionPValue(f: number, df1: number, df2: number): number {
  if (Number.isNaN(f)) return Number.NaN;
  if (f === Number.POSITIVE_INFINITY) return 0;
  if (f <= 0 || f === Number.NEGATIVE_INFINITY) return 1;
  const x = (df1 * f) / (df1 * f + df2);
  const p = regularizedBetaIncomplete(x, df1 / 2, df2 / 2);
  return clamp01(1 - p);
}

/** Two-sided Student's t p-value via the incomplete beta function. */
export function tDistributionPValue(t: number, df: number): number {
  if (df <= 0) return 1;
  if (Number.isNaN(t)) return Number.NaN;
  if (!Number.isFinite(t)) return 0;
  const x = df / (df + t * t);
  const p = regularizedBetaIncomplete(x, df / 2, 0.5);
  return clamp01(p);
}

/** Two-sided Student's t critical value at significance level alpha (bisection). */
export function tDistributionCriticalValue(alpha: number, df: number): number {
  if (alpha <= 0 || alpha >= 1) return Number.NaN;
  if (df <= 0) return Number.NaN;
  let lo = 0;
  let hi = 50;
  while (tDistributionPValue(hi, df) > alpha) {
    hi *= 2;
    if (hi > 1e5) return hi;
  }
  for (let i = 0; i < 120; i += 1) {
    const mid = (lo + hi) / 2;
    const p = tDistributionPValue(mid, df);
    if (p < alpha) {
      hi = mid;
    } else {
      lo = mid;
    }
    if (hi - lo < 1e-12) break;
  }
  return (lo + hi) / 2;
}

/** Inverse standard normal CDF (Acklam's rational approximation). */
export function inverseNormal(p: number): number {
  if (p <= 0) return Number.NEGATIVE_INFINITY;
  if (p >= 1) return Number.POSITIVE_INFINITY;

  const A = [
    -3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2, 1.38357751867269e2,
    -3.066479806614716e1, 2.506628277459239,
  ];
  const B = [
    -5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2, 6.680131188771972e1,
    -1.328068155288572e1,
  ];
  const C = [
    -7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838, -2.549732539343734,
    4.374664141464968, 2.938163982698783,
  ];
  const D = [7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996, 3.754408661907416];

  const plow = 0.02425;
  const phigh = 1 - plow;
  let approximation: number;
  if (p < plow) {
    const q = Math.sqrt(-2 * Math.log(p));
    approximation = (
      (((((C[0]! * q + C[1]!) * q + C[2]!) * q + C[3]!) * q + C[4]!) * q + C[5]!) /
      ((((D[0]! * q + D[1]!) * q + D[2]!) * q + D[3]!) * q + 1)
    );
  } else if (p > phigh) {
    const q = Math.sqrt(-2 * Math.log(1 - p));
    approximation = (
      -(((((C[0]! * q + C[1]!) * q + C[2]!) * q + C[3]!) * q + C[4]!) * q + C[5]!) /
      ((((D[0]! * q + D[1]!) * q + D[2]!) * q + D[3]!) * q + 1)
    );
  } else {
    const q = p - 0.5;
    const r = q * q;
    approximation = (
      ((((((A[0]! * r + A[1]!) * r + A[2]!) * r + A[3]!) * r + A[4]!) * r + A[5]!) * q) /
      (((((B[0]! * r + B[1]!) * r + B[2]!) * r + B[3]!) * r + B[4]!) * r + 1)
    );
  }

  // Acklam's approximation is accurate to about 1e-9. One Halley step,
  // using the incomplete-gamma identity for Phi(x), reaches double precision.
  const gammaArgument = (approximation * approximation) / 2;
  let cdfError: number;
  if (gammaArgument > 1.5) {
    const tail = regularizedUpperGammaCf(0.5, gammaArgument) / 2;
    cdfError = approximation < 0 ? tail - p : (1 - p) - tail;
  } else {
    const gamma = regularizedLowerGamma(0.5, gammaArgument);
    const cdf = approximation >= 0 ? 0.5 + gamma / 2 : 0.5 - gamma / 2;
    cdfError = cdf - p;
  }
  const density = Math.exp((-approximation * approximation) / 2) / SQRT_2PI;
  const correction = cdfError / density;
  return Number.isFinite(correction)
    ? approximation - correction / (1 + (approximation * correction) / 2)
    : approximation;
}

/** Two-sided upper-tail p-value for Student's t (named helper). */
export function studentTTwoSided(t: number, df: number): number {
  return tDistributionPValue(t, df);
}

/**
 * CDF of the noncentral t distribution, P(T ≤ t | df, ncp).
 *
 * Algorithm AS 243 (Lenth, 1989), as implemented in R's `pnt.c`: a
 * Poisson-weighted mixture of regularized incomplete beta terms. Used by the
 * power family (PROC POWER's noncentral-t sample size). Reduces to the central
 * t CDF when ncp ≈ 0.
 */
export function noncentralTCdf(t: number, df: number, ncp: number): number {
  if (!Number.isFinite(t) || df <= 0) return Number.NaN;
  if (Math.abs(ncp) < 1e-12) {
    const upperTwoSided = tDistributionPValue(t, df);
    return t >= 0 ? 1 - upperTwoSided / 2 : upperTwoSided / 2;
  }

  const errMax = 1e-12;
  const itrMax = 1000;
  const sqrt2dPi = Math.sqrt(2 / Math.PI);

  let negdel = false;
  let tt = t;
  let del = ncp;
  if (t < 0) {
    negdel = true;
    tt = -t;
    del = -ncp;
  }

  // x = tt^2 / (tt^2 + df); the central-normal part carries the rest.
  const x = (tt * tt) / (tt * tt + df);
  let tnc = 0;

  if (x > 0) {
    const lambda = del * del;
    let p = 0.5 * Math.exp(-0.5 * lambda);
    let q = sqrt2dPi * p * del;
    let s = 0.5 - p;
    const a = 0.5;
    const b = 0.5 * df;
    const rxb = Math.pow(1 - x, b);
    const albeta = logGamma(a) + logGamma(b) - logGamma(a + b);
    let xodd = regularizedIncompleteBeta(a, b, x);
    let godd = 2 * rxb * Math.exp(a * Math.log(x) - albeta);
    let xeven = 1 - rxb;
    let geven = b * x * rxb;
    tnc = p * xodd + q * xeven;

    let aa = a;
    for (let it = 1; it <= itrMax; it += 1) {
      aa += 1;
      xodd -= godd;
      xeven -= geven;
      godd *= (x * (aa + b - 1)) / aa;
      geven *= (x * (aa + b - 0.5)) / (aa + 0.5);
      p *= lambda / (2 * it);
      q *= lambda / (2 * it + 1);
      s -= p;
      tnc += p * xodd + q * xeven;
      const errbd = 2 * s * (xodd - godd);
      if (Math.abs(errbd) < errMax) break;
    }
  }

  tnc += normalCdf(-del);
  const result = negdel ? 1 - tnc : tnc;
  return clamp01(result);
}

export { regularizedIncompleteBeta };
